### ISSUE TITLE
DNS doesn't work anymore in VPN after update

### ISSUE THREAD

**Issue by ltuikov** [systemd/systemd#22658]
**systemd version the issue has been seen with**
 > systemd 249 (v249.9-1.fc35)
+PAM +AUDIT +SELINUX -APPARMOR +IMA +SMACK +SECCOMP +GCRYPT +GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 +PWQUALITY +P11KIT +QRENCODE +BZIP2 +LZ4 +XZ +ZLIB +ZSTD +XKBCOMMON +UTMP +SYSVINIT default-hierarchy=unified

**Used distribution**
 > Fedora 35

**Linux kernel version used** (`uname -a`)

 > `Linux localhost.localdomain 5.16.11-200.fc35.x86_64 #1 SMP PREEMPT Wed Feb 23 17:08:49 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux`

**CPU architecture issue was seen on**
 > AMD Ryzen

**Expected behaviour you didn't see**
 > DNS to work in VPN--it used to work BEFORE a package update I did a few weeks ago. (sudo dnf update)

**Unexpected behaviour you saw**
 > DNS cannot resolve a work website after connecting on VPN. It used to work fine before a package update a rew weeks ago.

**Steps to reproduce the problem**
 > Update to the latest packages from Fedora 35, run VPN to my work, try to reach a work website--can't resolve. But if I remote "search ." from /run/systemd/resolve/stub-resolv.conf and /run/systemd/resolve/resolv.conf, it works fine. The problem is "search .".

**Additional program output to the terminal or log subsystem illustrating the issue**

```text
...
```
The problem seems to be the unconditional addition of "search ." in resolv.conf by commit b3ffa2b5f3aa68.
"search ." should be added conditionally, say by a config option in /etc/systemd/resolved.conf, but not unconditionally. Or the correct fix should be implemented as described in the last paragraph in commit b3ffa2b5f3aa68, but definitely not adding "search ." unconditionally.
After connecting to work VPN, this line becomes "search . somthing.com" and this fails all DNS queries to work sites, and this used to work before a recent upgrade.

**Comment by poettering** [comment]
What VPN software do you use?

Does it patch /etc/resolv.conf itself? 

resolved only adds "search ." if no other domains are configured, it should never happen that it shows up in combnation with anything else, unless your VPN software writes resolv.conf directly, and does this incorrectly. Which it realy shouldn't if resolved is used, and definitely no naively, leaving "." at first place.

Either way, this looks like a bug in your VPN software, not resolved.

**Comment by ltuikov** [comment]
Hi,
I'm using openconnect(8) which uses vpnc(8), i.e. /etc/vpnc/vpnc-script to modify /etc/resolv.conf.
After it connects to the VPN, it leaves the search line as:
<pre>search . amd.com</pre>
Only after removing the '.' does it work correctly and work websites are being resolved.
So, yes, it does seem like a bug in vpnc, as /etc/resolv.conf says on the 1st line:
<pre>$head -1 /etc/resolv.conf 
#@VPNC_GENERATED@ -- this file is generated by vpnc
</pre>
So yes, this seems like a bug in vpnc-script.

**Comment by bluca** [comment]
Can you check if https://gitlab.com/openconnect/vpnc-scripts/-/commit/8f289ea712d8ce4f3d57acabe0813225501d0118 fixes the issue?

**Comment by ltuikov** [comment]
Hi Luca,

So I compared with other Fedora 35 up-to-date systems (just like the one I'd been using), and noticed that on them, vpnc does NOT update /etc/resolv.conf. Another strange thing I noticed at the same time DNS stopped working while on VPN is that after logging in, Wayland crashed and Gnome started instead and NetworkManager was down--so I had to restart it every time after logging in. So it seems like a snowball effect.

When it was all said and done, I reinstalled F35, salvaging /home, and at the moment things seem to work fine, i.e. F35 up-to-date and Wayland starts fine, and DNS in VPN to work sites works again, and vpnc DOES NOT update /etc/resolv.conf anymore (symlinked to the stub, as it's always been). So it was really strange what had caused this whole mess. FWIW, I wasn't sure what was causing vpnc-script to edit /etc/resolv.conf, and now on a reinstalled F35 up-to-date system, vpnc-script doesn't touch /etc/resolv.conf.

In the logs I could see that NetworkManager was getting SIGSEGV and complaining that "the hostname was changed from outside. Anyway, it is working now, fingers crossed. Should I notice a problem again, I'll update this thread.

Thanks for the pointer to the vpnc commit.

Regards,
Luben
